### ISSUE TITLE
[社区任务7] 基于Megatron+sglang完成MiMo-7B-的MTP训练

### ISSUE THREAD

**Issue by wucong25** [verl-project/verl-ascend-recipe#20]
- 任务描述
"昇腾上基于VeRL开放仓库的sglang推理后端，补齐该模型 MTP结构对于megatron后端的支持。
本期任务的具体信息如下：
（1）模型名称与算法：MiMo-7B + MTP
（2）VeRL仓库的模型源码链接：
https://github.com/verl-project/verl
（3）参考脚本：examples\mtp_trainer\run_mimo_7b_mtp_fully_async_megatron_multinode.sh
examples\mtp_trainer\run_mimo_7b_mtp_megatron.sh
（4） 开发平台：Atlas 800T A2或者A3"

- 验收标准
"""一、任务交付件
本期任务为基于verl开放仓库代码进行算法扩展，请合入开发代码。主要开发点如下：
（1） 模型适配patch
（2） 算法优化实现
二、验收标准：
1）精度/性能要求
精度：
长跑12小时或者训练100步，reward上升，平均误差小于5%，下游评测误差5%内。如果无GPU标杆，reward上升即可。
性能：
A2上达成A100的0.7倍以上。如果无GPU标杆，TPS绝对值大于100即可。
2）实践文档：
算法适配调优文档1篇
3）任务完成标准
本次任务完成标准为：
精度/性能（根据实际要求）达标，PR完成合入，实践文档提交到仓库issue。"""

- 任务类型
模型适配

**Comment by RordChang** [comment]
认领这个任务

**Comment by chengminhua** [comment]
> 认领这个任务

欢迎认领任务，请参考https://github.com/verl-project/verl/issues/6439 社区任务池明确该任务的：

完成的截止日期
开发进展反馈
微信答疑群
任务交付注意事项
等信息。如果您同时认领了多项任务，但无法都能进行投入，可以在部分任务中回复退出.

麻烦您加入到对应微信群，群备注名修改为"社区任务+您的gitcode账号", 后续有相关消息和问题都可以在微信群咨询答疑。 等您加入到微信群后，我这边会在社区任务池里面登记任务责任人。

**Comment by jianzhnie** [comment]
认领这个任务

**Comment by RordChang** [comment]
> > 认领这个任务
> 
> 欢迎认领任务，请参考[verl-project/verl#6439](https://github.com/verl-project/verl/issues/6439) 社区任务池明确该任务的：
> 
> 完成的截止日期 开发进展反馈 微信答疑群 任务交付注意事项 等信息。如果您同时认领了多项任务，但无法都能进行投入，可以在部分任务中回复退出.
> 
> 麻烦您加入到对应微信群，群备注名修改为"社区任务+您的gitcode账号", 后续有相关消息和问题都可以在微信群咨询答疑。 等您加入到微信群后，我这边会在社区任务池里面登记任务责任人。

您好，云开发平台提供的算力不足以支持训练，是否有解决方案

**Comment by chengminhua** [comment]
> > > 认领这个任务
> > 
> > 
> > 欢迎认领任务，请参考[verl-project/verl#6439](https://github.com/verl-project/verl/issues/6439) 社区任务池明确该任务的：
> > 完成的截止日期 开发进展反馈 微信答疑群 任务交付注意事项 等信息。如果您同时认领了多项任务，但无法都能进行投入，可以在部分任务中回复退出.
> > 麻烦您加入到对应微信群，群备注名修改为"社区任务+您的gitcode账号", 后续有相关消息和问题都可以在微信群咨询答疑。 等您加入到微信群后，我这边会在社区任务池里面登记任务责任人。
> 
> 您好，云开发平台提供的算力不足以支持训练，是否有解决方案

你好，已收到相关问题反馈，我们将尽快给出相关问题答复

**Comment by chengminhua** [comment]
> 认领这个任务

欢迎参加认领社区任务，当前一个任务只分配一个开发，根据评论认领时间确定，您可以后续观察该任务是否被释放退出，或者认领其他任务 https://gitcode.com/org/Ascend/discussions/4

**Comment by jianzhnie** [comment]
当前微信群过期了，能更新一下吗

**Comment by chengminhua** [comment]
> 当前微信群过期了，能更新一下吗

收到，正在联系相关负责人更新

**Comment by RordChang** [comment]
本周进展：已完成代码编写和相关配置，等待测试结果

**Comment by RordChang** [comment]
训练任务已完成，正在整理文档

**Comment by RordChang** [comment]
## MiMo-7B MTP 适配设计与复现结果

### 1. 任务目标

在 Atlas 800T A2 上完成 MiMo-7B-RL 的 MTP 训练，完整组合为：

- actor/reference：Megatron + MindSpeed + MBridge
- rollout：SGLang
- 权重同步：HCCL
- 设备：4 x Ascend 910B3
- 训练长度：100 steps

对应PR：[verl-project/verl-project/verl-ascend-recipe #87](https://github.com/verl-project/verl-ascend-recipe/pull/87)

### 2. 实现设计

本次在 verl 中新增注册式 `sglang_hccl` checkpoint engine，用于将 Megatron actor 权重直接同步到 colocated SGLang replicas。

权重同步流程：

1. 使用循环错位映射，为每个 Megatron rank 选择位于另一张卡上的 SGLang replica。
2. 为每个 trainer/rollout pair 建立持久化的两 rank HCCL process group。
3. 将 Hugging Face 格式的完整 tensors 按 bucket 分组，单个 tensor 不拆分。
4. 调用 SGLang 的 `init_weights_update_group` 初始化更新组。
5. 通过 `update_weights_from_distributed` 将参数直接发送到 SGLang。
6. 同步完成后恢复 SGLang memory/generation 状态；启用参数 offload 时，将 actor 参数重新放回 CPU。

MTP 训练配置：

```text
actor_rollout_ref.model.mtp.enable=True
actor_rollout_ref.model.mtp.enable_train=True
actor_rollout_ref.model.mtp.enable_rollout=False
actor_rollout_ref.model.mtp.mtp_loss_scaling_factor=0.1
actor_rollout_ref.model.mtp.detach_encoder=True
actor_rollout_ref.rollout.name=sglang
actor_rollout_ref.rollout.checkpoint_engine.backend=sglang_hccl
model_engine=megatron
```

本任务验证 MTP auxiliary loss 的训练。SGLang 负责 rollout，本配置不启用 rollout 阶段的 MTP acceleration。

### 3. 主要代码改动

- `verl/checkpoint_engine/sglang_hccl_checkpoint_engine.py`
  - HCCL process-group 初始化
  - Megatron/SGLang rank 配对
  - 完整 tensor bucket 化
  - distributed weight update
  - memory/generation 状态管理
- `verl/workers/engine_workers.py`
  - checkpoint-engine 同步完成后的参数 re-offload
- `examples/mtp_trainer/run_mimo_7b_mtp_megatron.sh`
  - 增加 `DEVICE=npu` 的 Ascend + SGLang 配置
  - 保留原有 CUDA 默认行为
- CPU tests
  - bucket 边界和顺序
  - 超大 tensor 报错
  - rank 错位映射
  - HTTP error/null response
  - 四卡 HCCL topology
  - global-step 传递
  - 参数 re-offload

### 4. 模型、数据与运行命令

模型使用 `XiaomiMiMo/MiMo-7B-RL`。下载后将模型 `config.json` 中的：

```json
{
  "max_position_embeddings": 32768
}
```

数据准备为 verl PPO Parquet 格式，并放置为：

```text
/path/to/math/train.parquet
/path/to/math/test.parquet
```

运行命令：

```bash
DEVICE=npu \
NPUS_PER_NODE=4 \
MODEL_PATH=/path/to/MiMo-7B-RL \
DATA_ROOT=/path/to/math \
bash examples/mtp_trainer/run_mimo_7b_mtp_megatron.sh
```

Ascend 默认配置：

- train/validation batch size：16/16
- PPO mini batch size：8
- micro batch size：1
- prompt/response length：512/1024
- Megatron TP/PP：2/1
- SGLang TP：1
- rollout n：2
- graph-mode max batch size：8
- weight bucket：1280 MiB
- total training steps：100
- save frequency：100
- resume mode：auto

### 5. 验证环境

- Atlas 800T A2
- 4 x Ascend 910B3，64 GiB HBM/NPU
- CANN 9.0.0
- Python 3.11.15
- torch 2.8.0
- torch-npu 2.8.0.post2
- SGLang 0.5.10
- sgl-kernel-npu 2026.2.1
- Megatron-Core/MindSpeed 0.16.0
- MBridge 0.15.1
- Transformers 5.3.0
- Ray 2.53.0

### 6. 训练结果

连续完成 100/100 steps，MTP loss 的 100 个样本均为有限值。

| 指标 | 结果 |
|---|---:|
| 前 10 步 reward 均值 | -0.8080 |
| 后 10 步 reward 均值 | 0.0891 |
| Reward 变化 | +0.8971 |
| Reward 线性斜率 | +0.01224/step |
| 前 10 步 MTP loss 均值 | 0.5979 |
| 后 10 步 MTP loss 均值 | 0.3559 |
| MTP loss 变化 | -0.2420 |
| 四卡端到端吞吐 | 533.3 token/s |
| 单 NPU 端到端吞吐 | 133.3 token/s/NPU |
| 稳态四卡吞吐 | 801.9 token/s |
| median step time | 40.18 s |

第 100 步中，每个 SGLang replica 接收：

- 451 个 tensors
- 12 个 buckets
- 15,666,819,072 bytes
- 同步时间 1.322-1.373 s

checkpoint tracker 正确保存为 step 100。训练 reward 上升、MTP loss 下降，端到端吞吐超过任务在无 GPU 标杆情况下的 TPS 要求。

### 7. 代码与检查

- PR：[verl-project/verl-ascend-recipe#87](https://github.com/verl-project/verl-ascend-recipe/pull/87)
- CPU tests：`12 passed, 1 warning`
- 完整 pre-commit：全部通过
- PR title checker：通过
- example naming checker：通过
- `bash -n`：通过

### 8. 完整训练日志

- [100-step 完整训练日志（已脱敏）](https://gist.github.com/RordChang/c7730e4b733b544105b1efaa14dab32b)

当前训练、核心适配代码、测试和复现材料均已提交，等待 PR review 与合入。
